docs/source/notes/extending.func.rst

```diff
@@ -0,0 +1,479 @@
+.. _func-autograd-function:
+
+Extending torch.func with autograd.Function
+===========================================
+
+.. currentmodule:: torch.autograd
+
+So you'd like to use :class:`torch.autograd.Function` with the :mod:`torch.func`
+transforms like :func:`torch.vmap`, :func:`torch.func.grad`, etc.
+
+There are two main use cases:
+
+- you wish to call code that does not contain PyTorch operations and
+  have it work with function transforms. That is, the :class:`torch.autograd.Function`'s
+  forward/backward/etc calls into functions from other systems like C++, CUDA, numpy.
+- you wish to specify custom gradient rules, like
+  JAX's `custom_vjp/custom_jvp <https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html>`_
```

Comment:
Do we want to mention that there might be unexpected behavior if the user tries to mix the two use cases. For example if user specifies a custom vmap rule AND wants to do custom gradients, the former can override the latter.

Reply:
Is that unexpected behavior though?

If I'm a user, and I write a forward() + backward() using only PyTorch operations, and I want custom vmap behavior, then I would write a custom vmap rule rather than use generate_vmap_rule.

Though I agree we should mention that this case is possible. I'll put it below in the "torch.vmap Support" section.

Reply:
yeah nah, not unexpected behavior, but probably tricky enough to deserve a mention

Reply:
Added a mention, as well as a small note about the grad(vmap(Func)) semantics